File: src/AIWordPressManager.Web/Services/BackgroundExecutionIdentity.cs
namespace AIWordPressManager.Web.Services;

/// <summary>
/// Carries the owning application user through one background execution flow.
/// The lease is ambient only for the current async execution context and must be disposed.
/// It never carries roles or administrator privileges.
/// </summary>
public static class BackgroundExecutionIdentity
{
    private static readonly AsyncLocal<IdentityState?> Current = new();

    public static bool TryGetOwnerUserId(out Guid ownerUserId)
    {
        var state = Current.Value;
        if (state is not null && state.OwnerUserId != Guid.Empty)
        {
            ownerUserId = state.OwnerUserId;
            return true;
        }

        ownerUserId = Guid.Empty;
        return false;
    }

    public static IDisposable Push(Guid ownerUserId)
    {
        if (ownerUserId == Guid.Empty)
            throw new ArgumentException("Background execution owner user ID is required.", nameof(ownerUserId));

        var previous = Current.Value;
        Current.Value = new IdentityState(ownerUserId);
        return new IdentityLease(previous);
    }

    private sealed record IdentityState(Guid OwnerUserId);

    private sealed class IdentityLease(IdentityState? previous) : IDisposable
    {
        private IdentityState? _previous = previous;
        private int _disposed;

        public void Dispose()
        {
            if (Interlocked.Exchange(ref _disposed, 1) != 0)
                return;

            Current.Value = _previous;
            _previous = null;
        }
    }
}

File: variants/laravel-aiwmweb/backend/app/Jobs/BackgroundExecutionIdentity.php
<?php

namespace App\Jobs;

use Closure;
use InvalidArgumentException;

/**
 * Scoped Laravel adaptation of the canonical BackgroundExecutionIdentity.
 *
 * Canonical operation: AIMW-AUTO-7E70C7119A.
 *
 * The identity deliberately carries only the owning Laravel user primary key.
 * It never authenticates that user and never carries roles, permissions or
 * administrator state. AppServiceProvider binds this service as scoped so a
 * long-running queue worker cannot leak an owner into the next job scope.
 */
final class BackgroundExecutionIdentity
{
    public const OPERATION_ID = 'AIMW-AUTO-7E70C7119A';

    private ?int $ownerUserId = null;

    public function tryGetOwnerUserId(): ?int
    {
        return $this->ownerUserId;
    }

    /**
     * Push an owner for the current execution scope and return an idempotent
     * lease that restores the previous owner when disposed.
     */
    public function push(int $ownerUserId): BackgroundExecutionIdentityLease
    {
        if ($ownerUserId < 1) {
            throw new InvalidArgumentException('Background execution owner user ID is required.');
        }

        $previous = $this->ownerUserId;
        $this->ownerUserId = $ownerUserId;

        return new BackgroundExecutionIdentityLease(function () use ($previous): void {
            $this->ownerUserId = $previous;
        });
    }
}

/**
 * Disposable lease for BackgroundExecutionIdentity::push().
 *
 * Explicit dispose mirrors the source IDisposable contract. The destructor is
 * a fail-safe for exceptional PHP control flow; dispose remains idempotent.
 */
final class BackgroundExecutionIdentityLease
{
    private bool $disposed = false;

    public function __construct(private ?Closure $restore) {}

    public function dispose(): void
    {
        if ($this->disposed) {
            return;
        }

        $this->disposed = true;
        $restore = $this->restore;
        $this->restore = null;
        $restore?->__invoke();
    }

    public function __destruct()
    {
        $this->dispose();
    }
}
